ssofact.module

```diff
@@ -35,23 +35,216 @@ function ssofact_help($route_name, RouteMatchInterface $route_match) {
 function ssofact_openid_connect_user_properties_ignore_alter(array &$properties_ignore, array $context) {
   unset($properties_ignore[array_search('mail', $properties_ignore)]);
   // @todo Triggers (logged) error: 'property type not implemented: changed'
-  //unset($properties_ignore[array_search('changed', $properties_ignore)]);
+  // unset($properties_ignore[array_search('changed', $properties_ignore)]);
 }
 
 /**
  * Implements hook_openid_connect_userinfo_save().
  */
 function ssofact_openid_connect_userinfo_save(UserInterface $account, $context) {
-  $userinfo = $context['userinfo'];
-  if (empty($userinfo['deactivated'])) {
+  $user_info = $context['userinfo'];
+  if (empty($user_info['deactivated'])) {
     $account->activate();
   }
   else {
     $account->block();
   }
 
   // Trigger event to let 3rd party to subscribe and alter.
-  \Drupal::service('event_dispatcher')->dispatch(SyncUserDataEvent::SYNC, new SyncUserDataEvent($account, $userinfo));
+  \Drupal::service('event_dispatcher')->dispatch(SyncUserDataEvent::SYNC, new SyncUserDataEvent($account, $user_info));
+
+  $article_test = \Drupal::service('role_paywall_article_test.manager');
+  $user_data = \Drupal::service('user.data');
+  $uid = $account->id();
+  if (isset($user_info['alfa_purchases']['purchases']) && is_array($user_info['alfa_purchases']['purchases'])) {
+    $purchases = $user_info['alfa_purchases']['purchases'];
+    $today = date('Ymd');
+    $active_purchases = [];
+    $expired_purchases = [];
+    $have_subscriptions = FALSE;
+    $subscription_timestamp = 0;
+    $have_free_trials = FALSE;
+    $free_trial_timestamp = 0;
+    $have_paid_trials = FALSE;
+    $paid_trial_timestamp = 0;
+    $product_labels = [];
+    $customer_status = '';
+    $subscription_product = '';
+    $last_purchase_timestamp = '';
+
+    // Classify purchases.
+    foreach ($purchases as $access) {
+      $purchase = $access['purchase'];
+      if (!is_product($purchase) && !is_free_trial($purchase) && !is_paid_trial($purchase)) {
+        continue;
+      }
+      // Assuming here no subscriptions schedules for the future.
+      if ($purchase['fromDay'] <= $today && $purchase['toDay'] >= $today) {
+        $active_purchases[] = $access['purchase'];
+      }
+      else {
+        $expired_purchases[] = $access['purchase'];
+      }
+    }
+
+    // No active subscriptions but article test active.
+    if (empty($active_purchases) && !$article_test->hasUserAccessToNextTest($account)) {
+      $customer_status = 'trial | article';
```

Comment:
Didn't we discuss to store the _"columns"_ of the defined tracking values as separate user data rows?

I did not expect to see pipes (`|`) in the user data values.  Those are the column separators.  The values of the columns should come from different user data rows.

The expected user data rows / values use only one column value, as we discussed and documented in the Asana ticket:
```
customer_status: trial
trial_access: article
subscription_product: ''
last_purchase_timestamp: 123456789
```
The tracking module code then concatenates those values to get the desired tracking value, along the lines of the following:
```php
$status = implode(' | ', array_filter([$customer_status, $trial_access, $subscription_product]));
$age = floor(($time - $last_purchase_timestamp) / 86400);
```

Storing the final tracking values instead of the atomic meanings makes the stored values not reusable for anything else (because we cannot hardcode the special meanings of tracking parameters with weird pipes in them into other places).